Ignore malformed environment variables on Windows too

tbu- opened the pull request:
See also #29297.

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented on src/libstd/sys/windows/os.rs:
```diff
@@ -89,12 +89,18 @@ impl Iterator for Env {
-            };
-            Some((OsStringExt::from_wide(k), OsStringExt::from_wide(v)))
-        }
+            if s.is_empty() {
```
This check doesn't need to be here as it was covered above via `if *self.cur == 0`, that's what indicates the end of the environment variable block.

alexcrichton commented on src/libstd/sys/windows/os.rs:
```diff
@@ -78,7 +78,7 @@ impl Iterator for Env {
 
     fn next(&mut self) -> Option<(OsString, OsString)> {
-        unsafe {
+        loop { unsafe {
```
Can you indent this as it would be normally?

tbu- commented:
I was under the impression that I read about that leading equals sign from official documentation of Microsoft, but I can't find it anymore. :/

tbu- commented:
@alexcrichton What should we do with malformed strings? There's nothing meaningful we can output. Upon encountering a string like `FOO`, we could either output `key="FOO", value=""`, `key="", value="FOO"`, `key="BAR", value="123"`, all of which are equally wrong. I think the best way to handle malformed strings is to ignore them.

tbu- commented:
Also, I think my parsing already treats a leading `=` as part of the variable name, I copied that from the UNIX counterpart.

alexcrichton commented:
I'm not really sure what the best strategy here is, I don't know what the system is actually generating. For example I'm not sure if it's even possible to see a string like `FOO` or perhaps even " FOO" (leading space). For now we don't have much documentation one way or another so I think it's best to just handle only the leading `=` case specially and just preserve some non-panicking behavior for the rest.

tbu- commented:
So what do you suggest to do in the "missing equals sign" case? The previous code special-cased it to put the whole thing into the key. My code special-cases it to skip the string.

alexcrichton commented:
I guess ignoring it isn't so bad yeah

tbu- commented:
@alexcrichton Addressed your comments.

alexcrichton commented on src/libstd/sys/windows/os.rs:
```diff
@@ -78,22 +78,28 @@ impl Iterator for Env {
+                self.cur = self.cur.offset(len + 1);
+
+                // `s` has at least length 1 at this point, because the empty
+                // string terminates the array of environment variables.
```
Can you update this comment to mention that the purpose of this slice-after-one is not because we know it has length one but because we want to skip any leading `=` character?

alexcrichton commented:
I think this can re-un-ignore the test in run-pass as well, right?

tbu- commented:
Done.

alexcrichton commented:
@bors: r+ e1740221461089106612fd45ffad80f8805abba6

bors commented:
:hourglass: Testing commit e174022 with merge 4ef5e2a...

bors commented:
:broken_heart: Test failed - [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7113)

tbu- commented:
Fixed the error, compiled it locally (didn't know you could uncover compilation errors in a Windows module on Linux), squashed the commits.

This might take a few test cycles to complete, I don't have a Windows machine where I can test the build.

alexcrichton commented:
@bors: r+ 9b4f16b370c2c4b44de99960a2bf7300c93e3f9e

bors commented:
:hourglass: Testing commit 9b4f16b with merge 6861c51...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7163), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7177), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/436), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7116), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7138), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7123), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/353), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2303), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7195), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7189), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7154), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2146), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2150), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2151), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2145), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1405), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2133)
